### ISSUE TITLE
Invalid shares after connection reset

### ISSUE THREAD

**Issue by sammy007** [str4d/zcash#14]
If pool resets connections (restart) miner (and zogminer) will reconnect and all shares will be invalid on pool side.

**Comment by mmitech** [comment]
I can reproduce this as well

**Comment by sammy007** [comment]
Related because code is the same https://github.com/nginnever/zogminer/issues/65

**Comment by str4d** [comment]
My guess is that after the reconnect, there's an assumption being made either by my Stratum code or the pool about what data is persisted and what is regenerated. Most likely, the cause is that the server provides a new nonce1 to the miner, but the miner is not seeing a new job and so continues mining with the old nonce1. I haven't yet reviewed my code though, and I'd need to see a connection trace to be sure.

**Comment by sammy007** [comment]
New connection on pool side, so new nonce is generated and sent on `mining.subscribe` followed by a target and job after `mining.authorize`. Session ID is `null` of course.

**Comment by str4d** [comment]
Hmm, then the miner *should* get notified about the new job and switch to it. I'll look at this later today.

**Comment by sammy007** [comment]
This is very necessary to have reference implementation 100% robust and compliant, we are suppering from this mess already,f or example zogminer borrowed this stratum code and already ignores `notify` after `set_target` after `authorize` and mining inly starts when new job pushed from server (on new block in network).

**Comment by sammy007** [comment]
I sniffed traffic on port, hope it helps:

```
T 192.168.1.2:9998 -> 192.168.1.4:44040 [AP]
  {"id":null,"method":"mining.notify","params":["699187805179202" ... }
##
T 192.168.1.4:44040 -> 192.168.1.2:9998 [A]
  {"id": 4, "method": "mining.submit", "params": ["X","699187805179202" ... }
##
T 192.168.1.2:9998 -> 192.168.1.4:44040 [AP]
  {"id":4,"result":true,"error":null}.

######## STRATUM RESET ########

T 192.168.1.2:9998 -> 192.168.1.4:44042 [AR]
  ......
####
T 192.168.1.4:44044 -> 192.168.1.2:9998 [AP]
  {"id": 1, "method": "mining.subscribe", "params": ["/MagicBean:1.0.0/", null,"192.168.1.2","9998"]}.
###
T 192.168.1.2:9998 -> 192.168.1.4:44044 [AP]
  {"id":1,"result":[null,"000000000000000000000000051638b6"],"error":null}.
##
T 192.168.1.4:44044 -> 192.168.1.2:9998 [AP]
  {"id": 2, "method": "mining.authorize", "params": ["X","x"]}.
##
T 192.168.1.2:9998 -> 192.168.1.4:44044 [AP]
  {"id":2,"result":true,"error":null}.
#
T 192.168.1.2:9998 -> 192.168.1.4:44044 [AP]
  {"id":null,"method":"mining.set_target","params":["015555555555555555555555555555555555555555555555555555555555552a"]}.
##
T 192.168.1.2:9998 -> 192.168.1.4:44044 [AP]
  {"id":null,"method":"mining.notify","params":["762369330492885","04000000","...","...","00000000000 ... }
  00000000000000000000000000000000000000000000000000000","7c511b58","4af6031d",true]}.
{"id": 4, "method": "mining.submit", "params": ["X","762369330492885","7c511b58" ... }
##
T 192.168.1.2:9998 -> 192.168.1.4:44044 [AP]
  {"id":4,"result":null,"error":[23,"Low difficulty share",null]}.
```

Nonce is new after pool reset, it's randomised for sharding.

**Comment by str4d** [comment]
Looks like you are using my standalone miner for testing. Could you add `-debug=stratum` to its CLI and then post the submission output? You should be able to see the full nonce (including server portion) when it makes a submission.

**Comment by sammy007** [comment]
I added `-debug=stratum -printtoconsole`

```
Solution found; Submitting to 192.168.1.2...
  d5555555555555555555555555555561b9381605000000000000000000000000
B-) Submitted and accepted.
[WARN] read_until: End of file
Reconnecting in 3 seconds...
Connecting to stratum server 192.168.1.2:9998
[ERROR] Could not connect to stratum server 192.168.1.2:9998, Connection refused
Reconnecting in 3 seconds...
[WARN] read_until: Transport endpoint is not connected
Reconnecting in 3 seconds...
Connecting to stratum server 192.168.1.2:9998
Connected!
Subscribed to stratum server
Authorized worker X
Target set to 015555555555555555555555555555555555555555555555555555555555552a
Received new job #371186289322138
Found solution satisfying the server target
Solution found; Submitting to 192.168.1.2...
  55555555555555555555555555555594b9381605000000000000000000000000
Low difficulty share
[WARN] :-( Not accepted.
Found solution satisfying the server target
Solution found; Submitting to 192.168.1.2...
  55555555555555555555555555555594b9381605000000000000000000000000
Low difficulty share
[WARN] :-( Not accepted.
```

We can see it clearly that old part of nonce is used.

BTW, miner still mine even if server is down, I see that with `-debug=pow`

**Comment by str4d** [comment]
@sammy007 could you please apply the following patch to my standalone miner and then test it again:

```diff
diff --git a/src/libstratum/ZcashStratum.cpp b/src/libstratum/ZcashStratum.cpp
index 7315fcc..2042d30 100644
--- a/src/libstratum/ZcashStratum.cpp
+++ b/src/libstratum/ZcashStratum.cpp
@@ -319,12 +319,14 @@ void ZcashMiner::stop()
 void ZcashMiner::setServerNonce(const Array& params)
 {
     auto n1str = params[1].get_str();
+    LogPrintf("New nonce1 received from server: \"%s\"\n", n1str);
     std::vector<unsigned char> nonceData(ParseHex(n1str));
     while (nonceData.size() < 32) {
         nonceData.push_back(0);
     }
     CDataStream ss(nonceData, SER_NETWORK, PROTOCOL_VERSION);
     ss >> nonce1;
+    LogPrintf("Updated nonce template to %s\n", nonce1.GetHex());
 
     nonce1Size = n1str.size();
     size_t nonce1Bits = nonce1Size * 4; // Hex length to bit length
@@ -387,6 +389,8 @@ ZcashJob* ZcashMiner::parseJob(const Array& params)
     ret->nonce2Space = nonce2Space;
     ret->nonce2Inc = nonce2Inc;
 
+    LogPrintf("Applied nonce template %s to job %s\n", nonce1.GetHex(), ret->job);
+
     return ret;
 }
 
```

**Comment by sammy007** [comment]
```
Solution found; Submitting to 192.168.1.2...
  bffffffffffffffffffffffffffffffc4789b601000000000000000000000000
B-) Submitted and accepted.
[WARN] read_until: End of file
Reconnecting in 3 seconds...
Connecting to stratum server 192.168.1.2:9998
Connected!
Subscribed to stratum server
New nonce1 received from server: "000000000000000000000000034c09b9"
Updated nonce template to 00000000000000000000000000000000b9094c03000000000000000000000000
Authorized worker X
Target set to 015555555555555555555555555555555555555555555555555555555555552a
Applied nonce template 00000000000000000000000000000000b9094c03000000000000000000000000 to job 114231506221785
Received new job #114231506221785
Found solution satisfying the server target
Solution found; Submitting to 192.168.1.2...
  4000000000000000000000000000000b4789b601000000000000000000000000
Low difficulty share
[WARN] :-( Not accepted.
Found solution satisfying the server target
Solution found; Submitting to 192.168.1.2...
  800000000000000000000000000000094789b601000000000000000000000000
Low difficulty share
[WARN] :-( Not accepted.
```

tl;dr new nonce ***applied*** but stale nonce still used.

**Comment by str4d** [comment]
This is useful, thanks. The confusing thing here is that you're setting `CLEAN_JOBS` to true, which means that the miners should cancel their current work and pick up the new job immediately. But even after a few stales, the new job should be detected. Time to go bug-hunting...

**Comment by daira** [comment]
Moved to https://github.com/zcash/zcash/issues/2381 ; should be closed here.
